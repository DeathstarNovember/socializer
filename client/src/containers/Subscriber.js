import { useEffect } from "react";

const Subscriber = ({ subscribeToNew, children }) => {
  useEffect(() => {
    subscribeToNew();
  }, [subscribeToNew]);

  return children;
};

export default Subscriber;
