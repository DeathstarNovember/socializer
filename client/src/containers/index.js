export { default as StateProvider } from "./StateProvider";
export { default as Subscriber } from "./Subscriber";
