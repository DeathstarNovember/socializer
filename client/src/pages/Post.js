import React, { useCallback } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { Col, Container, Row } from "react-bootstrap";
import { Helmet } from "react-helmet";
import produce from "immer";
import { ChatBar, Feed, FeedItem, QueryResult } from "components";

export const GET_POST = gql`
  query GetPost($id: String!) {
    post(id: $id) {
      id
      body
      insertedAt
      user {
        id
        name
        gravatarMd5
      }
      comments {
        id
        body
        insertedAt
        user {
          id
          name
          gravatarMd5
        }
      }
    }
  }
`;

export const COMMENTS_SUBSCRIPTION = gql`
  subscription onCommentCreated($postId: String!) {
    commentCreated(postId: $postId) {
      id
      body
      insertedAt
      user {
        id
        name
        gravatarMd5
      }
    }
  }
`;

const Post = ({
  match: {
    params: { id },
  },
}) => {
  const { subscribeToMore, ...queryResult } = useQuery(GET_POST, {
    variables: { id },
  });

  const subscribeToNew = useCallback(
    () =>
      subscribeToMore({
        document: COMMENTS_SUBSCRIPTION,
        variables: { postId: id },
        updateQuery: (prev, { subscriptionData }) => {
          if (!subscriptionData.data) return prev;
          const newComment = subscriptionData.data.commentCreated;

          // Check that we don't already have
          // the comment stored.
          if (
            prev.post.comments.find((comment) => comment.id === newComment.id)
          ) {
            return prev;
          }

          return produce(prev, (next) => {
            next.post.comments.push(newComment);
          });
        },
      }),
    [id],
  );

  return (
    <>
      <Helmet>
        <title>Socializer | Discussion</title>
        <meta property="og:title" content="Socializer | Discussion" />
      </Helmet>
      <Container>
        <QueryResult {...queryResult}>
          {({ data }) => (
            <Row>
              <Col md={4} className="d-none d-md-block">
                <ChatBar />
              </Col>
              <Col xs={12} md={8}>
                <h4>Discussion</h4>
                <FeedItem item={data.post} />
                <hr />
                <h5>Comments</h5>
                <Feed
                  feedType="comment"
                  newItemPosition="bottom"
                  items={data.post.comments}
                  createParams={{ postId: id }}
                  subscribeToNew={subscribeToNew}
                />
              </Col>
            </Row>
          )}
        </QueryResult>
      </Container>
    </>
  );
};

export default Post;
