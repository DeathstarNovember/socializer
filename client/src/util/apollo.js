import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
