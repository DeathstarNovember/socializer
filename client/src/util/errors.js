import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
