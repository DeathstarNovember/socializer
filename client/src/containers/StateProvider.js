import React, { useEffect, useState } from "react";
import { useApolloClient } from "@apollo/react-hooks";
import Cookies from "js-cookie";
import { refreshSocket } from "util/apollo";
import { AuthContext, ChatContext } from "util/context";

const StateProvider = ({ initialToken, initialUserId, socket, children }) => {
  const client = useApolloClient();
  const [token, setToken] = useState(initialToken || Cookies.get("token"));
  const [userId, setUserId] = useState(initialUserId || Cookies.get("userId"));
  const [chatState, setChatState] = useState("default");

  // If the token changed (i.e. the user logged in
  // or out), clear the Apollo store and refresh the
  // websocket connection.
  useEffect(() => {
    if (!token) client.clearStore();
    if (socket) refreshSocket(socket);
  }, [client, socket, token]);

  const setAuth = (data) => {
    if (data) {
      const { id, token } = data;
      Cookies.set("token", token);
      Cookies.set("userId", id);
      setToken(token);
      setUserId(id);
    } else {
      Cookies.remove("token");
      Cookies.remove("userId");
      setToken(null);
      setUserId(null);
    }
  };

  return (
    <AuthContext.Provider value={{ token, userId, setAuth }}>
      <ChatContext.Provider value={{ chatState, setChatState }}>
        {children}
      </ChatContext.Provider>
    </AuthContext.Provider>
  );
};

export default StateProvider;
