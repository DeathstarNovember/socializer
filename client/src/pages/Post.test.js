import React from "react";
import { render, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import tk from "timekeeper";
import { Subscriber } from "containers";
import { AuthContext, ChatContext } from "util/context";
import Post, { GET_POST, COMMENTS_SUBSCRIPTION } from "./Post";

jest.mock("components/ChatBar", () => () => <div className="chat-bar" />);

jest.mock("containers/Subscriber", () =>
  jest.fn().mockImplementation(({ children }) => children),
);

const getMock = {
  request: {
    query: GET_POST,
    variables: { id: 1 },
  },
  result: {
    data: {
      post: {
        id: 1,
        body: "Here are a few thoughts",
        insertedAt: "2019-04-17T16:00:00",
        user: {
          id: 1,
          name: "Steven Paul",
          gravatarMd5: "abc",
        },
        comments: [
          {
            id: 1,
            body: "I agree!",
            insertedAt: "2019-04-17T18:00:00",
            user: {
              id: 2,
              name: "Patrick Lane",
              gravatarMd5: "def",
            },
          },
        ],
      },
    },
  },
};

describe("Post", () => {
  beforeEach(() => {
    tk.freeze("2019-04-20");
  });

  afterEach(() => {
    tk.reset();
  });

  it("renders correctly when loading", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ userId: 1 }}>
          <ChatContext.Provider value={{ chatState: "default" }}>
            <MockedProvider mocks={[getMock]} addTypename={false}>
              <Post match={{ params: { id: 1 } }} />
            </MockedProvider>
          </ChatContext.Provider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when loaded", async () => {
    const { container, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ userId: 1 }}>
          <ChatContext.Provider value={{ chatState: "default" }}>
            <MockedProvider mocks={[getMock]} addTypename={false}>
              <Post match={{ params: { id: 1 } }} />
            </MockedProvider>
          </ChatContext.Provider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );

    await wait(() => getByText("Here are a few thoughts"));
    expect(container).toMatchSnapshot();
  });

  it("renders correctly after created comment", async () => {
    Subscriber.mockImplementation((props) => {
      const { default: ActualSubscriber } = jest.requireActual(
        "containers/Subscriber",
      );
      return <ActualSubscriber {...props} />;
    });

    const mocks = [
      getMock,
      {
        request: {
          query: COMMENTS_SUBSCRIPTION,
          variables: { postId: 1 },
        },
        result: {
          data: {
            commentCreated: {
              id: 2,
              body: "Thanks for your support!",
              insertedAt: "2019-04-17T19:00:00",
              user: {
                id: 1,
                name: "Steven Paul",
                gravatarMd5: "abc",
              },
            },
          },
        },
      },
    ];
    const { container, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ userId: 1 }}>
          <ChatContext.Provider value={{ chatState: "default" }}>
            <MockedProvider mocks={mocks} addTypename={false}>
              <Post match={{ params: { id: 1 } }} />
            </MockedProvider>
          </ChatContext.Provider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    await wait(() => getByText("Thanks for your support!"));
    expect(container).toMatchSnapshot();
  });
});
