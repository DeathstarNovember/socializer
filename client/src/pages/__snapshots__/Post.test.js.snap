// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post renders correctly after created comment 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="d-none d-md-block col-md-4"
      >
        <div
          class="chat-bar"
        />
      </div>
      <div
        class="col-md-8 col-12"
      >
        <h4>
          Discussion
        </h4>
        <div
          class="mb-4 card"
        >
          <div
            class="d-flex justify-content-between align-items-center card-header"
          >
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Gravatar for undefined"
                class="react-gravatar rounded-circle"
                height="30"
                src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
                srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
                width="30"
              />
              <div
                class="ml-2"
              >
                <div
                  class="h6 mb-0"
                >
                  Steven Paul
                </div>
              </div>
            </div>
            <div
              class="text-muted h7"
            >
              <time
                datetime="2019-04-17T16:00:00.000Z"
                title="2019-04-17T16:00:00Z"
              >
                2 days ago
              </time>
            </div>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              Here are a few thoughts
            </p>
          </div>
        </div>
        <hr />
        <h5>
          Comments
        </h5>
        <div
          class="mb-4 card"
        >
          <div
            class="d-flex justify-content-between align-items-center card-header"
          >
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Gravatar for undefined"
                class="react-gravatar rounded-circle"
                height="30"
                src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
                srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
                width="30"
              />
              <div
                class="ml-2"
              >
                <div
                  class="h6 mb-0"
                >
                  Patrick Lane
                </div>
              </div>
            </div>
            <div
              class="text-muted h7"
            >
              <time
                datetime="2019-04-17T18:00:00.000Z"
                title="2019-04-17T18:00:00Z"
              >
                2 days ago
              </time>
            </div>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              I agree!
            </p>
          </div>
        </div>
        <div
          class="mb-4 card"
        >
          <div
            class="d-flex justify-content-between align-items-center card-header"
          >
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Gravatar for undefined"
                class="react-gravatar rounded-circle"
                height="30"
                src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
                srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
                width="30"
              />
              <div
                class="ml-2"
              >
                <div
                  class="h6 mb-0"
                >
                  Steven Paul
                </div>
              </div>
            </div>
            <div
              class="text-muted h7"
            >
              <time
                datetime="2019-04-17T19:00:00.000Z"
                title="2019-04-17T19:00:00Z"
              >
                2 days ago
              </time>
            </div>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              Thanks for your support!
            </p>
          </div>
        </div>
        <div
          class="mb-4 card"
        >
          <div
            class="card-body"
          >
            <div
              class="text-muted h7"
            >
              <a
                href="/login"
              >
                Log in
              </a>
               to submit a 
              comment
              .
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Post renders correctly when loaded 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="d-none d-md-block col-md-4"
      >
        <div
          class="chat-bar"
        />
      </div>
      <div
        class="col-md-8 col-12"
      >
        <h4>
          Discussion
        </h4>
        <div
          class="mb-4 card"
        >
          <div
            class="d-flex justify-content-between align-items-center card-header"
          >
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Gravatar for undefined"
                class="react-gravatar rounded-circle"
                height="30"
                src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
                srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
                width="30"
              />
              <div
                class="ml-2"
              >
                <div
                  class="h6 mb-0"
                >
                  Steven Paul
                </div>
              </div>
            </div>
            <div
              class="text-muted h7"
            >
              <time
                datetime="2019-04-17T16:00:00.000Z"
                title="2019-04-17T16:00:00Z"
              >
                2 days ago
              </time>
            </div>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              Here are a few thoughts
            </p>
          </div>
        </div>
        <hr />
        <h5>
          Comments
        </h5>
        <div
          class="mb-4 card"
        >
          <div
            class="d-flex justify-content-between align-items-center card-header"
          >
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Gravatar for undefined"
                class="react-gravatar rounded-circle"
                height="30"
                src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
                srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
                width="30"
              />
              <div
                class="ml-2"
              >
                <div
                  class="h6 mb-0"
                >
                  Patrick Lane
                </div>
              </div>
            </div>
            <div
              class="text-muted h7"
            >
              <time
                datetime="2019-04-17T18:00:00.000Z"
                title="2019-04-17T18:00:00Z"
              >
                2 days ago
              </time>
            </div>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              I agree!
            </p>
          </div>
        </div>
        <div
          class="mb-4 card"
        >
          <div
            class="card-body"
          >
            <div
              class="text-muted h7"
            >
              <a
                href="/login"
              >
                Log in
              </a>
               to submit a 
              comment
              .
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Post renders correctly when loading 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="d-flex justify-content-center"
    >
      <div
        class="spinner-border"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
